cluster/tls_connection.go

```diff
@@ -0,0 +1,170 @@
+// Copyright 2020 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cluster
+
+import (
+	"bufio"
+	"crypto/tls"
+	"encoding/binary"
+	"io"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/hashicorp/memberlist"
+	"github.com/pkg/errors"
+
+	"github.com/prometheus/alertmanager/cluster/clusterpb"
+)
+
+const (
+	version      = "v1.0.0"
+	uint32length = 4
+)
+
+// tlsConn wraps net.Conn with connection pooling data.
+type tlsConn struct {
+	connection net.Conn
+	timeout    time.Duration
+	lock       sync.Mutex
```

Comment:
nit: throughout a couple of files here. We tend to use the name of `mtx`  for the locks and place it above the variables it's protecting followed by a space. I don't think it's consistent throughout _all_ of the codebase but is never a bad time to start.

```suggestion
	mtx       sync.Mutex
```